docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,230 @@
+  state: ... (Terraform state)
+- kind: ControlPlane
+  purpose: normal
+  state: ... (Certificates generated by the extension)
+- kind: Worker
+  state: ... (Machine objects)
+```
+
+The state data is saved as a `runtime.RawExtension` type, which can be encoded/decoded by the corresponding extension controller.
+
+There can be sensitive data in the `ShootState` which has to be hidden from the end-users. Hence, it will be recommended to provide an etcd encryption configuration to the Gardener API server in order to encrypt the `ShootState` resource.
+
+#### Size limitations
+
+There are limits on the size of the request bodies sent to the kubernetes API server when creating or updating resources: by default ETCD can only accept request bodies which do not exceed 1.5 MiB (this can be configured with the `--max-request-bytes` flag); the kubernetes API Server has a request body limit of 3 MiB which cannot be set from the outside (with a command line flag); the gRPC configuration used by the API server to talk to ETCD has a limit of 2 MiB per request body which cannot be configured from the outside; and `watch` requests have a 16 MiB limit on the buffer used to stream resources.
+
+This means that if `ShootState` is bigger than 1.5 MiB, the ETCD max request bytes will have to be increased. However, there is still an upper limit of 2 MiB imposed by the gRPC configuration.
+
+If `ShootState` exceeds this size limitation it must make use of configmap/secret references to store the state of extension controllers. This is an implementation detail of Gardener and can be done at a later time if necessary as extensions will not be affected.
+
+Splitting the `ShootState` into multiple resources could have a positive benefit on performance as the Gardener API Server and Gardener Controller Manager would handle multiple small resources instead of one big resource.
+
+### Gardener extensions changes
+
+All extension controllers which require state migration must save their state in a new `status.state` field and act on an annotation `gardener.cloud/operation=restore` in the respective Custom Resources which should trigger a restoration operation instead of reconciliation. A restoration operation means that the extension has to restore its state in the Shoot's namespace on the __Destination Seed__ from the `status.state` field.
+
+As an example: the `Infrastructure` resource must save the terraform state.
+
+```
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: Infrastructure
+metadata:
+  name: infrastructure
+  namespace: shoot--foo--bar
+spec:
+  type: azure
+  region: eu-west-1
+  secretRef:
+    name: cloudprovider
+    namespace: shoot--foo--bar
+  providerConfig:
+    apiVersion: azure.provider.extensions.gardener.cloud/v1alpha1
+    kind: InfrastructureConfig
+    resourceGroup:
+      name: mygroup
+    networks:
+      vnet: # specify either 'name' or 'cidr'
+      # name: my-vnet
+        cidr: 10.250.0.0/16
+      workers: 10.250.0.0/19
```

Comment:
Do we need to capture VM information here?

Reply:
What do you mean? 
These fields are present in the `Shoot.spec.provider.infrastructureConfig`

Reply:
What I mean is do we need a field like `Shoot.spec.provider.infrastructureConfig.virtualMachines`? I was trying to figure out if we need to capture the list of VMs running on the provider?

Reply:
I don't understand. Can you please elaborate more what you are trying to achieve, @prashanth26 ?

Reply:
Okay, I was under the assumption that we might have to migrate the VMs from one account to another, hence was wondering if we need to transfer VM state. But later I realized that's not the case. 

So you can ignore my comment and resolve this issue.